ops/model.py

```diff
@@ -3613,69 +3603,25 @@ def __init__(
-        token = self._is_recursive.set(True)
+    def _trace_and_wrap_errors(self, cmd: str, *args: Any, **kwargs: Any):
         try:
-            yield
-        finally:
-            self._is_recursive.reset(token)
-
-    def _run(
-        self,
-        *args: str,
-        return_output: bool = False,
-        use_json: bool = False,
-        input_stream: str | None = None,
-    ) -> str | Any | None:
-        if self._is_recursive.get():
-            # Either `juju-log` hook command failed or there's a bug in ops.
-            return
-        # Logs are collected via log integration, omit the subprocess calls that push
-        # the same content to juju from telemetry.
-        mgr = (
-            self._prevent_recursion()
-            if args[0] == 'juju-log'
-            else tracer.start_as_current_span(args[0])
-        )
-        with mgr as span:
-            kwargs = {
-                'stdout': subprocess.PIPE,
-                'stderr': subprocess.PIPE,
-                'check': True,
-                'encoding': 'utf-8',
-            }
-            if input_stream:
-                kwargs.update({'input': input_stream})
-            which_cmd = shutil.which(args[0])
-            if which_cmd is None:
-                raise RuntimeError(f'command not found: {args[0]}')
-            args = (which_cmd,) + args[1:]
-            if use_json:
-                args += ('--format=json',)
-            if span:
+            with tracer.start_as_current_span(cmd) as span:
                 span.set_attribute('call', 'subprocess.run')
-                # The command-line arguments for some hook commands may include sensitive data.
-                truncate = args[0] in ['action-set']
-                span.set_attribute('argv', [args[0], '...'] if truncate else args)
-            # TODO(benhoyt): all the "type: ignore"s below kinda suck, but I've
-            #                been fighting with Pyright for half an hour now...
-            try:
-                result = subprocess.run(args, **kwargs)  # type: ignore
-            except subprocess.CalledProcessError as e:
-                self._check_for_security_event(args[0], e.returncode, e.stderr)
-                raise ModelError(e.stderr) from e
-            if return_output:
-                if result.stdout is None:  # type: ignore
-                    return ''
-                else:
-                    text: str = result.stdout  # type: ignore
-                    if use_json:
-                        return json.loads(text)  # type: ignore
-                    else:
-                        return text  # type: ignore
+                span.set_attribute('argv', args)
+                span.set_attribute('kwargs', [f'{k}={v}' for k, v in kwargs.items()])
+                yield span
+        except hookcmds.Error as e:
+            self._check_for_security_event(e.cmd[0], e.returncode, e.stderr)
+            if (
+                cmd.startswith(('relation-', 'network-'))
```

Comment:
Ah, nice -- I like how you've done this with specific tests here, rather than special modes/keyword args.